### ISSUE TITLE
CQUI_RealHousingFromImprovements still necessary / can we share code?

### ISSUE THREAD

**Issue by the-m4a** [civfanatics/CQUI_Community-Edition#34]
CQUI_RealHousingFromImprovements calculates the housing value for each city from the improvements around it, which required work when the Maya pack was released.  The base game did this calculation without apparent issue.  CQUI_RealHousingFromImprovements is something we may not need to maintain any longer and just let the base game do its thing.

However IF we do, then we DO have similar code in 3 places (vanilla, exp1, exp2) that feels like it doesn't need to be repeated like this.  the code itself isn't the most clear, either...

For a simple example: 
```
            local kImprovementData = GameInfo.Improvements[eImprovementType].Housing;
            if kImprovementData == 1 then    -- farms, pastures etc.
``` 
kImprovementData is the amount of housing that comes from that improvement type.  Note that except for the Maya, farms provide 0.5 housing, though it says 1 above.  This is because in the database (debugGameplay.sqllite), the housing values are all integers in the Improvements table.  

For what it's worth, there doesn't seem to be a consistent flow with finding these requirements:
for example, walking through finding the data for Kampung:

Table: Improvements
- Filter: IMPROVEMENT_KAMPUNG
- (all other values no meaning right now)

Table: ImprovementModifiers
- ImprovementType: IMPROVEMENT_KAMPUNG
- ModifierId: KAMPUNG_HOUSING_WITHTECH

Table: Modifiers
- ModifierId: KAMPUNG_HOUSING_WITHTECH
- SubjectRequirementSetId: PLAYER_HAS_KAMPUNG_TECHNOLOGY_REQUIREMENTS 

Table: RequirementSetRequirements
- RequirementSetId: PLAYER_HAS_KAMPUNG_TECHNOLOGY_REQUIREMENTS
- Value: REQUIRES_PLAYER_HAS_MASS_PRODUCTION_TECHNOLOGY

Table: Requirements 
- RequirementId: REQUIRES_PLAYER_HAS_MASS_PRODUCTION_TECHNOLOGY
- RequirementType:  REQUIREMENT_PLAYER_HAS_TECHNOLOGY
- ProgressWeight = 1
- All other values 0

It's unclear from there exactly how from there as there's no specific consistency in picking out that technology...

**Comment by Infixo** [comment]
1. The problem with the in-game GetHousingFromImprovements() function is that it rounds down the Housing. As example you can never know if you have 1.0 or 1.5 free, in both cases the UI will show 1. Thus you don't know if you need to plop 1 or 2 farms to unblock growth. This is the reason why CQUI counts it manually in the first place.
2. In the Improvements table there is also a field TilesRequired. It is usually 2 which corresponds to 0.5 from Farms, etc. However, to count correctly you need to be prepared that some mod may introduce let's say 2 housing from 3 improvements. It is possible. So you need to add up Housing/TilesRequired and not Housing and then at the end divide by 2.
3. Some Improvements gain extra housing with technologies or civics. This is always done via modifiers, so to account for that you need to get this info dynamically from the modifier definitions. To make things more confusing those amounts are always integers. You can't give 0.5 housing via a modifier.
4. Then there is Maya, which is another exception. Her farms give 1.5 housing (0.5 from the table, and 1 from the modifier).
5. I am not sure how is know, but last time I checked CQUI counted only worked tiles, but imo the tile needs to be improved to provide housing, no need to work it.
6. In Better Report Screen I wrote my own piece of code that calculates all of the above. Lines 291-365. You can use the code in CQUI, I am ok with that. However, you still have to code the modifier analysis because I am using RMA for that (Real Modifier Analysis), which is part of BRS.

**Comment by the-m4a** [comment]
Awesome, thank you for the info!  Very good to know.

**Comment by Infixo** [comment]
There may be an easier way. This topic got me thinking a bit and maybe it would be enough just to calculate the correction from improvements and apply it. This way you would not have to dabble in the modifiers. Just an idea, need thorough testing.

**Comment by Infixo** [comment]
Seems to be working.

```
-- 2020-06-09 Infixo: new idea for calculations - calculate only a correction and apply to the game function
local iCityMaxBuyPlotRange:number = tonumber(GlobalParameters.CITY_MAX_BUY_PLOT_RANGE);
function GetRealHousingFromImprovements(pCity:table)
	local cityX:number, cityY:number = pCity:GetX(), pCity:GetY();
	local iNumHousing:number = 0; -- we'll add data from Housing field in Improvements divided by TilesRequired which is usually 2
	-- check all plots in the city
	for _,plotIndex in ipairs(Map.GetCityPlots():GetPurchasedPlots(pCity)) do
		local pPlot:table = Map.GetPlotByIndex(plotIndex);
		if pPlot and pPlot:GetImprovementType() > -1 and not pPlot:IsImprovementPillaged() and Map.GetPlotDistance(cityX, cityY, pPlot:GetX(), pPlot:GetY()) <= iCityMaxBuyPlotRange then
			local imprInfo:table = GameInfo.Improvements[ pPlot:GetImprovementType() ];
			iNumHousing = iNumHousing + imprInfo.Housing / imprInfo.TilesRequired; -- well, we can always add 0, right?
		end
	end
	return pCity:GetGrowth():GetHousingFromImprovements() + Round(iNumHousing-math.floor(iNumHousing),1);
end
```

**Comment by the-m4a** [comment]
I'm not sure this code is working right - at least for the Maya.  I'm testing right now - just basic stuff to see if I needed to halve the value or not... so maybe I am missing something.

The code I have in-game:
```
-- at top of file
local CQUI_cityMaxBuyPlotRange :number = tonumber(GlobalParameters.CITY_MAX_BUY_PLOT_RANGE);
--
-- inside the calculate housing function:
    local cityX:number, cityY:number = pCity:GetX(), pCity:GetY();
    local iNumHousing:number = 0; -- we'll add data from Housing field in Improvements divided by TilesRequired which is usually 2
    -- check all plots in the city
    for _,plotIndex in ipairs(Map.GetCityPlots():GetPurchasedPlots(pCity)) do
        local pPlot:table = Map.GetPlotByIndex(plotIndex);
        if (pPlot ~= nil 
            and pPlot:GetImprovementType() > -1
            and not pPlot:IsImprovementPillaged()
            and Map.GetPlotDistance(cityX, cityY, pPlot:GetX(), pPlot:GetY()) <= CQUI_cityMaxBuyPlotRange) then
            local imprInfo:table = GameInfo.Improvements[ pPlot:GetImprovementType() ];
            print_debug("!!!! iNumHousing: "..tostring(iNumHousing).."  imprInfo.Housing:"..tostring(imprInfo.Housing).."   TilesRequired:"..tostring(imprInfo.TilesRequired));
            iNumHousing = iNumHousing + (imprInfo.Housing / imprInfo.TilesRequired); -- well, we can always add 0, right?
        end
    end

    local baseHousingFromImprovements = pCity:GetGrowth():GetHousingFromImprovements();
    local realHousingValue = baseHousingFromImprovements + Round(iNumHousing - math.floor(iNumHousing), 1);
    print_debug("!!! BaseHousingFromImprovements: "..tostring(baseHousingFromImprovements));
    print_debug("!!! iNumHousing value: "..tostring(iNumHousing));
    print_debug("!!! Updated housing value: "..tostring(realHousingValue));
```

City - far enough away from other things that I shouldn't have any interference.  2 farms only at the moment.

![image](https://user-images.githubusercontent.com/8787640/85217476-79287a80-b346-11ea-8b6a-0e052051e8e7.png)

I'd expect the housing from improvements value to be 2.  The existing CQUI code calculates it (4, i have the print statement there before the divide by 2)

**Comment by the-m4a** [comment]
Adding a Seastead, it's better-ish:

![image](https://user-images.githubusercontent.com/8787640/85217669-467f8180-b348-11ea-9e23-ac3b087f30da.png)

**Comment by Infixo** [comment]
I think this incident was solved together with seasteads #42 and can be closed.

**Comment by the-m4a** [comment]
Yes indeed, closing
